File: problem-solving/_lc-427-medium-construct-quad-tree/1-solution.py
"""
https://leetcode.com/problems/construct-quad-tree/

Complexity
----------
Time: O(n^2)
Space: O(logn)  // We are dividing by n by half in every call.

Reference: https://leetcode.com/problems/construct-quad-tree/solutions/154565/java-recursive-solution/
"""

"""
# Definition for a QuadTree node.
class Node:
    def __init__(self, val, isLeaf, topLeft, topRight, bottomLeft, bottomRight):
        self.val = val
        self.isLeaf = isLeaf
        self.topLeft = topLeft
        self.topRight = topRight
        self.bottomLeft = bottomLeft
        self.bottomRight = bottomRight
"""

class Solution:
    def construct(self, grid: List[List[int]]) -> 'Node':
        def qTree(r, c, n):
            """(r, c) is coordinates of top left."""
            if n == 1:
                return Node(grid[r][c] != 0, True, None, None, None, None)
            h = n // 2
            newNode = Node()
            topLeft = qTree(r, c, h)
            topRight = qTree(r, c + h, h)
            bottomLeft = qTree(r + h, c, h)
            bottomRight = qTree(r + h, c + h, h)
            children = [topLeft, topRight, bottomLeft, bottomRight]
            # All are leaf and have same val
            if all(x.isLeaf and x.val == topLeft.val for x in children):
                newNode.val = topLeft.val
                newNode.isLeaf = True
            else:
                newNode.topLeft = topLeft
                newNode.topRight = topRight
                newNode.bottomLeft = bottomLeft
                newNode.bottomRight = bottomRight
                # newNode.val can be True or False
            return newNode
        return qTree(0, 0, len(grid))

File: problem-solving/_lc-427-medium-construct-quad-tree/1-solution.go
/*
https://leetcode.com/problems/construct-quad-tree/

Complexity
----------
Time: O(n^2)
Space: O(logn)  // We are dividing by n by half in every call.

Reference: https://leetcode.com/problems/construct-quad-tree/solutions/154565/java-recursive-solution/
*/

/**
 * Definition for a QuadTree node.
 * type Node struct {
 *     Val bool
 *     IsLeaf bool
 *     TopLeft *Node
 *     TopRight *Node
 *     BottomLeft *Node
 *     BottomRight *Node
 * }
 */

func construct(grid [][]int) *Node {
    var F func(int, int, int) *Node
    // (r, c) is coordinates of top left.
    F = func(r, c, n int) *Node {
        if n == 1 {
            return &Node{
                Val: grid[r][c] != 0,
                IsLeaf: true,
                TopLeft: nil,
                TopRight: nil,
                BottomLeft: nil,
                BottomRight: nil,
            }
        }
        node := &Node{}
        h := n / 2
        tLeft := F(r, c, h)
        tRight := F(r, c + h, h)
        bLeft := F(r + h, c, h)
        bRight := F(r + h, c + h, h)
        // All are leaf and have same val
        if tLeft.IsLeaf && tRight.IsLeaf && bLeft.IsLeaf && bRight.IsLeaf && tLeft.Val == tRight.Val && tRight.Val == bLeft.Val && bLeft.Val == bRight.Val {
            node.Val = tLeft.Val
            node.IsLeaf = true
        } else {
            node.TopLeft = tLeft
            node.TopRight = tRight
            node.BottomLeft = bLeft
            node.BottomRight = bRight
            // node.Val can be true or false
            // node.IsLeaf will be false by default
        }
        return node
    }
    return F(0, 0, len(grid))
}
